### PR TITLE
[Feature] Update ReasonReact JSX to version 3

### PR DESCRIPTION
**Pull request by arecvlohe**
## Summary

This PR updates the ReasonReact jsx syntax to version 3, bsb-platform to 5.0.4, bs-react-helmet to 2.0.0, and reason-react to 0.7.0. In addition, the example reason-react-app is now updated to use jsx version 3.

### CODE REVIEW

File: packages/plugin-renderer-react/src/link.re [hunk 1]
```diff
@@ -1,23 +1,13 @@
...
-    ~props={
-      "href": Js.Nullable.fromOption(href),
-      "style": Js.Nullable.fromOption(style),
-      "activeStyle": Js.Nullable.fromOption(activeStyle),
-      "className": Js.Nullable.fromOption(className),
-      "activeClassName": Js.Nullable.fromOption(activeClassName),
-    },
-    children,
-  );
+[@react.component]
```
Comment (arecvlohe) [inline packages/plugin-renderer-react/src/link.re]: I should probably revert this as this is not a `[@react.component]`

File: packages/plugin-renderer-react/src/phenomicRendererReactClient.re [hunk 2]
```diff
@@ -1,16 +1,16 @@
 [@bs.module "@phenomic/plugin-renderer-react/lib/client"]
 external createContainer_:
-  (ReasonReact.reactClass, Js.t({..})) => ReasonReact.reactClass =
+  (ReasonReact.reactClass, Js.t({..})) => React.element =
```
Comment (arecvlohe) [inline packages/plugin-renderer-react/src/phenomicRendererReactClient.re]: Neither are these.

File: yarn.lock [hunk 3]
```diff
@@ -2982,6 +2982,11 @@ bs-platform@^5.0.0:
   resolved "https://registry.yarnpkg.com/bs-platform/-/bs-platform-5.0.0.tgz#f9fded818bafc3891aeb85eb4e0d95b8d8f8b4d7"
   integrity sha512-zxLobdIaf/r7go47hOgxcd6C0ANh7MYMEtZNb9Al7JdoekzFZI50F4GpZpP8kMh9Z+M5PoPE1WuryT4S8mT8Kg==
 
+bs-platform@^5.0.3:
```
Comment (arecvlohe) [inline yarn.lock]: This is duped.

File: package.json [hunk 4]
```diff
@@ -147,5 +147,8 @@
     "website",
     "examples/*",
     "packages/*"
-  ]
+  ],
+  "dependencies": {
+    "reason-react": "^0.7.0"
+  }
```
Comment (MoOx) [inline package.json]: This isn't necessary.
You should add/modify this dep in appropriate package & run `yarn` at the root of the repo. Yarn _workspace_ feature handle all the necessary work to correctly install deps & make symlink if necessary.

File: package.json [hunk 5]
```diff
@@ -22,7 +22,7 @@
     "babel-core": "7.0.0-bridge.0",
     "babel-eslint": "^10.0.0",
     "babel-jest": "^23.6.0",
-    "bs-platform": "^5.0.0",
+    "bs-platform": "^5.0.3",
```
Comment (MoOx) [inline package.json]: Please update this deps in example/reason-react-app & in packages/bs-platform too, for consitency :)

File: packages/plugin-renderer-react/src/link.re (hunk 1 above)
Comment (MoOx) [reply packages/plugin-renderer-react/src/link.re]: This is a component, so this looks good.

File: packages/plugin-renderer-react/src/phenomicRendererReactClient.re (hunk 2 above)
Comment (MoOx) [reply packages/plugin-renderer-react/src/phenomicRendererReactClient.re]: Here indeed that's different this function take a component & return a component. So `React.component('a)` should be used instead reactClass

**Comment by MoOx** [comment]
To be sure the thing is working, you also need to edit example/reason-react-app.
That's where you can specify an actual reason-react version & run a build :)

Good start anyway!

**Comment by arecvlohe** [comment]
Cool! I will try and work on this more tonight, but might be tomorrow. Also, I realized that we might have to use `ReasonReactCompat.wrapReasonReactForReact` in cases where we can't convert jsx 2 to jsx 3.

**Comment by arecvlohe** [comment]
@MoOx Still just working along and making necessary changes but running into an issue that I am not able to diagnose:

```shell
sh: /Users/arecvlohe/Documents/development/phenomic/node_modules/bs-platform/lib/bsppx.exe -bs-jsx 3: No such file or directory
File "", line 1:
Error: Error while running external preprocessor
Command line: '/Users/arecvlohe/Documents/development/phenomic/node_modules/bs-platform/lib/bsppx.exe -bs-jsx 3' '/var/folders/h6/x5vgndfs0dv_y2017t23x51m0000gn/T/camlppx3bf7df' '/var/folders/h6/x5vgndfs0dv_y2017t23x51m0000gn/T/camlppxf075b9'
```

The error will show in your editor at the very top of `link.re`. It seems like everything else builds fine. I was having this issue in other files but it's only an issue in one now. I will take a look at this some more tomorrow night but wanted to leave it here in case you could take a look at it.

File: packages/plugin-renderer-react/src/phenomicRendererReactClient.re [hunk 6]
```diff
@@ -1,16 +1,14 @@
 [@bs.module "@phenomic/plugin-renderer-react/lib/client"]
-external createContainer_:
-  (ReasonReact.reactClass, Js.t({..})) => ReasonReact.reactClass =
+external createContainer_: (React.component('a), Js.t({..})) => React.element =
```
Comment (MoOx) [inline packages/plugin-renderer-react/src/phenomicRendererReactClient.re]: I think it's more returning the same React.component (kind of), & same below. Right?

Comment (arecvlohe) [reply packages/plugin-renderer-react/src/phenomicRendererReactClient.re]: What do you mean here?

**Comment by arecvlohe** [comment]
@MoOx Thanks for the context, it looks like the error is coming from RLS. I still have errors when running `yarn run test` so I will into that more today.

**Comment by arecvlohe** [comment]
@MoOx Leaving a comment here because I can't figure this out. When I run `yarn run transpile` I am getting a syntax error in `link.re` on the line where it has `React.element`. I am not sure why this is happening. I Also found it strange that in `Link.js` no children are passed but when it's used a `React.array` is passed to `<Link />`. Maybe this is shows a lack of understanding but wanted to make note of it so I can learn something. Anyhow, the syntax error is main thing that is blocking me so any help would be appreciated!

**Comment by arecvlohe** [comment]
@MoOx Sorry for spamming you on a Saturday but I think the code is now ready. After running `yarn run test`, all the tests pass. It looks as though there is a build error though, something related to using `&&`. You probably would know more about that than me. Anyhow, have a good weekend and let me know if the code needs any more tweaking!

File: examples/reason-react-app/App.js [hunk 7]
```diff
@@ -7,22 +7,16 @@ import { withPhenomicApi } from "@phenomic/preset-react-app/lib/es6/src/phenomic
 
 import * as Home from "./lib/es6/src/components/Home.bs.js";
 import * as Post from "./lib/es6/src/components/Post.bs.js";
-import ErrorPage from "./lib/es6/src/components/ErrorPage.bs.js";
+import { make as ErrorPage } from "./lib/es6/src/components/ErrorPage.bs.js";
```
Comment (MoOx) [inline examples/reason-react-app/App.js]: Maybe we could avoid that by doing `let default = make`?

Comment (arecvlohe) [reply examples/reason-react-app/App.js]: Yeah, that makes sense. I don't have a preference either way.

**Comment by arecvlohe** [comment]
@MoOx Done. I can see the benefit of splitting it up into different packages. Might make it easier to maintain overall. If more people contribute/maintain then it would be easier for one person to maintain one library instead of many.

File: packages/plugin-renderer-react/src/phenomicRendererReactClient.re (hunk 6 above)
Comment (MoOx) [reply packages/plugin-renderer-react/src/phenomicRendererReactClient.re]: Should this be `(React.component('a), Js.t({..})) => React.component('a)` ? an hoc isn't returning a rendered element, but almost the same component

**Comment by arecvlohe** [comment]
@MoOx Done. I thought this made more sense `(React.component('a), Js.t({..})) => React.component('b)` since it's a HOC, it should be going from '`a` to `'b`.

**Comment by arecvlohe** [comment]
@MoOx ping

**Comment by arecvlohe** [comment]
@MoOx Ping

**Comment by MoOx** [comment]
1.0.0 is out on npm (just include this change). Thanks for helping!

**Comment by arecvlohe** [comment]
@MoOx Awesome!

**Pull request merged**